docs/website/contents/for-developers/NodeTasks.md

```diff
@@ -0,0 +1,305 @@
+  PeerABlockFetch -- blocks --> ChainSelectionAndStorageLayer
+  PeerBBlockFetch -- blocks --> ChainSelectionAndStorageLayer
+
+  ChainSelectionAndStorageLayer -- headers --> PeerCChainSync
+  ChainSelectionAndStorageLayer -- blocks --> PeerCBlockFetch
+  ChainSelectionAndStorageLayer -- headers --> PeerDChainSync
+  ChainSelectionAndStorageLayer -- blocks --> PeerDBlockFetch
+```
+
+Here, BlockFetchLogic is a centralized component that decides whether and in which order to download certain blocks based on the candidate header chains.
+
+#### Diffusion of transactions
+
+```mermaid
+graph LR
+  subgraph PeerA[Upstream Peer A]
+    PeerATxSubmission[TxSubmission]
+  end
+
+  subgraph PeerB[Upstream Peer B]
+    PeerBTxSubmission[TxSubmission]
+  end
+
+  subgraph PeerC[Downstream Peer C]
+    PeerCTxSubmission[TxSubmission]
+  end
+
+  subgraph PeerD[Downstream Peer D]
+    PeerDTxSubmission[TxSubmission]
+  end
+
+  Mempool
+
+  Mempool --> PeerATxSubmission
+  Mempool --> PeerBTxSubmission
+  PeerCTxSubmission --> Mempool
+  PeerDTxSubmission --> Mempool
+```
+
+## Relaying a new block
+
+As a concrete example, we will consider how the node relays a new block B from its upstream peers to its downstream peers.
+We assume that no other blocks arrive in the meantime, and that B extends the node's selection.
+
+We also provide rough numbers on the time budget of the steps for illustration.
+
+ 1. In the ChainSync mini protocol thread for peer X, we receive a `MsgRollForward` with the header of B.
+    We validate the header and add it to the candidate header chain for peer X.
+
+    Time budget: average ~1ms, upper bound ~2ms.
+
+ 2. The BlockFetchLogic wakes up due to a change to a candidate chain, and notices the new header.
+    As the header extends our selection, the BlockFetchLogic tells the BlockFetch client for peer X to download the full block B.
+
+    Time budget: essentially instant.
+
+ 3. The BlockFetch client for peer X downloads the full block B and submits it to chain selection.
+
+    Time budget: depends on the bandwidth and latency between the node and peer X, as well as on the block size (at most 88 KiB at the moment). Usually, it should be (significantly) less than one second.
+
+ 4. Before chain selection starts, B is written to disk.
+
+    This is very cheap (no `fsync`).
+
+ 5. The node checks whether B is eligible for *block diffusion pipelining*, an optimization which allows to already announce a block to downstream peers before its body has been validated.
+    In scenarios like this one, this is usually the case.
+
+ 6. Because of block diffusion pipelining for B, the ChainSync servers to our downstream peers send a `MsgRollForward` containing the header of B.
+
+    Soon after, the downstream peers will request the corresponding block body of B via BlockFetch if they do not have it.
+
+ 7. In parallel, chain selection for B continues by validating B.
+
+    Time budget: average ~30ms, upper bound ~500ms.
```

Comment:
Where does `500ms` come from? EG the June attack exceeded this -- what's the scope being considered here? Perhaps it's none-buggy-block validation 😬

Reply:
Thanks for pointing that out, I chose that number somewhat arbitrarily. According to the recent db-analyser pass, there are only 1596 blocks with block application mutator time higher than 500ms, out of which 1574 (all but 22 blocks) are due to the ref script DoS.

I added a "based on empirical data" caveat.